pinot-plugins/pinot-stream-ingestion/pinot-kinesis/src/main/java/org/apache/pinot/plugin/stream/kinesis/KinesisConsumer.java

```diff
@@ -93,11 +93,31 @@ private KinesisMessageBatch getKinesisMessageBatch(KinesisPartitionGroupOffset s
       return new KinesisMessageBatch(List.of(), startOffset, true);
     }
 
-    // Read records
-    rateLimitRequests();
-    GetRecordsRequest getRecordRequest =
-        GetRecordsRequest.builder().shardIterator(shardIterator).limit(_config.getNumMaxRecordsToFetch()).build();
-    GetRecordsResponse getRecordsResponse = _kinesisClient.getRecords(getRecordRequest);
+    // Read records from kinesis.
+    // Based on getRecords documentation, we might get a response with empty records but a non-null nextShardIterator.
+    // This method is also used to accurately determine if we reached end of shard. So, we need to use nextShardIterator
+    // and call getRecords again until we get non-empty records or null nextShardIterator.
+    // To prevent an infinite loop due to some bug, we will limit the number of attempts
+    GetRecordsResponse getRecordsResponse;
+    int attempts = 0;
+    while (true) {
```

Comment:
Let's remove this. We had this while loop code a while back but chose to remove it in #12806 
we want the consumer thread to handle empty record response

Reply:
thanks. I see that primarily that change to remove the while loop was to address the "Return the message batch immediately without combining multiple of them" through removing "messages.size() >= _config.getNumMaxRecordsToFetch()"

This particular behaviour doesn't change and we won't combine multiple batches. 

The new while loop just handles cases where message batch is empty and shard iterator is not null. It iterates until it finds a null shard iterator or a non empty message batch. 

this seems like a good change right ? although we can make the client in the client method consumedEndOfShard() but I think adding it to getKinesisMessageBatch() is more appropriate so that the case is handled for other use cases too. 

what do you suggest now ?

Reply:
No, the reason we also did it was because it was leading to confusion when debugging via logs where consumer thread was retrying vs KinesisConsumer object.
I'd prefer keeping this the old way

Reply:
done